checks/raw/pinned_dependencies.go

```diff
@@ -61,9 +61,113 @@ func PinningDependencies(c *checker.CheckRequest) (checker.PinningDependenciesDa
 		return checker.PinningDependenciesData{}, err
 	}
 
+	if isStagedNugetDepsUnpinned(&results) {
+		if err := collectInsecureNugetCsproj(c, &results); err != nil {
+			return checker.PinningDependenciesData{}, err
+		}
+	} else {
+		promoteStagedNugetDependencies(&results, false)
+	}
 	return results, nil
 }
 
+func collectInsecureNugetCsproj(c *checker.CheckRequest, dependencies *checker.PinningDependenciesData) error {
+	csprojDeps, pinnedCsprojDeps, err := collectCsprojDependenciesData(c)
+	if err != nil {
+		return err
+	}
+	// some csproj files are pinned, output these to the user
+	if pinnedCsprojDeps > 0 && pinnedCsprojDeps <= len(csprojDeps) {
+		dependencies.Dependencies = append(dependencies.Dependencies, csprojDeps...)
+	} else {
+		allDependenciesArePinned := pinnedCsprojDeps == len(csprojDeps)
+		promoteStagedNugetDependencies(dependencies, allDependenciesArePinned)
+	}
+
+	return nil
+}
+
+func promoteStagedNugetDependencies(dependencies *checker.PinningDependenciesData, updateDependencyPinning bool) {
+	nugetDeps := dependencies.GetStagedDependencies(checker.DependencyUseTypeNugetCommand)
+
+	// all csproj files are pinned, negate the pinned status of all nuget dependencies
+	if updateDependencyPinning {
+		for i := 0; i < len(nugetDeps); i++ {
+			nugetDeps[i].Pinned = asBoolPointer(true)
+			nugetDeps[i].Remediation = nil
+		}
+	}
+
+	// add all NugetDependencies to Dependencies
+	dependencies.Dependencies = append(dependencies.Dependencies, nugetDeps...)
+}
+
+func collectCsprojDependenciesData(c *checker.CheckRequest) ([]checker.Dependency, int, error) {
+	var csprojDeps []checker.Dependency
+	if err := fileparser.OnMatchingFileContentDo(c.RepoClient, fileparser.PathMatcher{
+		Pattern:       "*.csproj",
+		CaseSensitive: false,
+	}, analyseCsprojLockedMode, &csprojDeps); err != nil {
+		return nil, 0, err
+	}
+
+	pinnedDependencies := countPinned(csprojDeps)
+	return csprojDeps, pinnedDependencies, nil
+}
+
+func analyseCsprojLockedMode(path string, content []byte, args ...interface{}) (bool, error) {
+	pdata, ok := args[0].(*[]checker.Dependency)
+	if !ok {
+		// panic if it is not correct type
+		panic(fmt.Sprintf("expected type []checker.Dependency, got %v", reflect.TypeOf(args[0])))
+	}
+
+	err, pinned := fileparser.IsRestoreLockedModeEnabled(content)
+	if err != nil {
+		return true, err
+	}
+
+	dependency := checker.Dependency{
+		Location: &checker.File{
+			Path:      path,
+			Type:      finding.FileTypeSource,
+			Offset:    1,
+			EndOffset: 1,
+			Snippet:   "hello",
```

Comment:
Is there a reason for this "hello"? leftover debug?

Reply:
fixed, thanks